src/Specular/Internal/Incremental.purs

```diff
@@ -0,0 +1,410 @@
+  let deps = [toSomeNode a]
+  runEffectFn1 Node.create 
+    { compute: mkEffectFn1 \_ -> do
+        value_a <- runEffectFn1 Node.get_value a
+        pure (if Optional.isSome value_a
+              then fn (Optional.fromSome value_a)
+              else Optional.none)
+    , dependencies: pure deps
+    }
+
+map2 :: forall a b c. EffectFn3 (Fn2 a b c) (Node a) (Node b) (Node c)
+map2 = mkEffectFn3 \fn a b -> do
+  let deps = [toSomeNode a, toSomeNode b]
+  runEffectFn1 Node.create 
+    { compute: mkEffectFn1 \_ -> do
+        value_a <- runEffectFn1 Node.valueExc a
+        value_b <- runEffectFn1 Node.valueExc b
+        pure (Optional.some (runFn2 fn value_a value_b))
+    , dependencies: pure deps
+    }
+
+-- Problem with bind and connect:
+-- We have to:
+-- - first compute LHS
+-- - then compute ourselves
+-- - only then we know the second dependency
+-- - compute the second dependency
+--
+-- Problem with bind and compute:
+-- - if we compute and the LHS changes, our height changes! We may need to re-add ourselves to recompute queue
+bind_ :: forall a b. EffectFn2 (Node a) (a -> Node b) (Node b)
+bind_ = mkEffectFn2 \lhs fn -> do
+  runEffectFn3 switch true lhs fn
+
+switch :: forall a b. EffectFn3 Boolean (Node a) (a -> Node b) (Node b)
+switch = mkEffectFn3 \alwaysFire lhs fn -> do
+  main_node_ref <- runEffectFn1 Ref.new Optional.none
+  rhs_node <- runEffectFn1 Node.create 
+    { compute: mkEffectFn1 \node -> do
+
+        value_lhs <- runEffectFn1 Node.valueExc lhs
+        let rhs = fn value_lhs
+
+        -- adjust dependencies and height of the main node, taking new dependency into account
+        main_opt <- runEffectFn1 Ref.read main_node_ref
+        let main = Optional.fromSome main_opt
+
+        runEffectFn2 addDependent rhs (toSomeNode main)
+        dependencyHeight <- runEffectFn1 Node.get_height rhs
+        runEffectFn2 ensureHeight main (dependencyHeight + 1)
+
+        -- disconnect from old dependency, if present
+        old_rhs_opt <- runEffectFn1 Node.get_value node
+        if Optional.isSome old_rhs_opt then do
+          runEffectFn2 removeDependent (Optional.fromSome old_rhs_opt) (toSomeNode main)
+        else pure unit
+
+        pure (Optional.some rhs)
+    , dependencies: do
+        pure [toSomeNode lhs]
+    }
+  runEffectFn2 Node.annotate rhs_node "switch aux"
+  main <- runEffectFn1 Node.create 
+    { compute: mkEffectFn1 \_ -> do
+        rhs <- runEffectFn1 Node.valueExc rhs_node
+        isFiring <- runEffectFn1 Node.isChangingInCurrentStabilization rhs
+        if alwaysFire || isFiring then do
+          runEffectFn1 Node.get_value rhs
+        else
+          pure Optional.none
+    , dependencies: do
+        rhs_opt <- runEffectFn1 Node.get_value rhs_node
+        if Optional.isSome rhs_opt then
+          pure [toSomeNode rhs_node, toSomeNode (Optional.fromSome rhs_opt)]
+        else
+          -- if we don't know the rhs yet, `compute` in rhs proxy will add it
```

Comment:
When can this happen?

Reply:
Maybe in the case of `height(main) < height(rhs) && adjustedHeight(main) > height(rhs)` (ie. if the main node will be re-added to the queue at a bigger height in stabilize).

Reply:
If we're `connect`ing the main (output) node for the first time, the control node has not been computed yet (because its existence is only discovered via `dependencies` of the output node).